src/dbnode/storage/bootstrap/bootstrapper/fs/source_index_test.go

```diff
@@ -526,131 +527,179 @@ func TestBootstrapIndexWithPersistPrefersPersistedIndexBlocks(t *testing.T) {
 // right now it only builds a partial segment for the second of three index
 // blocks it is trying to build.
 func TestBootstrapIndexWithPersistForIndexBlockAtRetentionEdge(t *testing.T) {
-	dir := createTempDir(t)
-	defer os.RemoveAll(dir)
-
-	timesOpts := testTimesOptions{
-		numBlocks: 3,
+	tests := []struct {
+		name                         string
+		now                          time.Time
+		expectedInfoFiles            int
+		expectedSegmentsFirstBlock   int
+		expectedSegmentsSecondBlock  int
+		expectedOutOfRetentionBlocks int64
+	}{
+		{
+			name:                         "now",
+			now:                          time.Now(),
+			expectedInfoFiles:            2,
+			expectedSegmentsFirstBlock:   1,
+			expectedSegmentsSecondBlock:  1,
+			expectedOutOfRetentionBlocks: 0,
+		},
+		{
+			name:                         "now + 4h",
+			now:                          time.Now().Add(time.Hour * 4),
+			expectedInfoFiles:            1,
+			expectedSegmentsFirstBlock:   0,
+			expectedSegmentsSecondBlock:  1,
+			expectedOutOfRetentionBlocks: 1,
+		},
+		{
+			name:                         "now + 8h",
+			now:                          time.Now().Add(time.Hour * 8),
+			expectedInfoFiles:            0,
+			expectedSegmentsFirstBlock:   0,
+			expectedSegmentsSecondBlock:  0,
+			expectedOutOfRetentionBlocks: 2,
+		},
 	}
-	times := newTestBootstrapIndexTimes(timesOpts)
-	firstIndexBlockStart := times.start.Truncate(testIndexBlockSize)
-
-	writeTSDBGoodTaggedSeriesDataFiles(t, dir, testNs1ID, times.start)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			dir := createTempDir(t)
```

Comment:
Might have been a good idea to extract the test implementation from test table loop into a dedicated private function. This would remove 2 levels of nesting and probably result in a much cleaner diff.

Reply:
Could you point me an example where such extraction is done? 
To me, anonymous method makes valuable role here by accessing a test context so there is no need to pass it or declare it.

Reply:
Usually it is not that useful, but in this case the body is 140 lines long, and the resulting diff is really hard to digest.
All of the context that you would have to pass would be just `(t, test)`.